### DISCUSSION TITLE
Create scripts to restrict participants functions and actions

### DISCUSSION THREAD

**Discussion by lgaleano07** [OpenTTD/OpenTTD#12936] [General]
hello!
i am new to teh game and I have zero knowledge on programming. But I want to use the gaem for my Phd project and Would like to manipulate/change an scenario I have created, so some members of the team are not able at some pint of having an entire view of the map.
i would also want to introduce accidents at a certain moment, always the same, and always the same magnitud, so I can compate teams.
is there anyway I can do this?

I was lloking for info on how to change or create a script, I haven´t found in a firts place how to create a new script, I don´t see the option in the general menu.

I Know these are many questions. I will appreciate your help.

Thanks!

**Comment by PikkaBird** [comment]
It sounds like gamescripts ( https://wiki.openttd.org/en/Manual/Game%20script ) are what you want to look into. Things like "not having an entire view of the map" may be more-or-less achievable in GS, but I suspect your actual goal might be better realised in other ways, once you have a better understanding of the game.

**Reply by lgaleano07** [reply]
thank you!
I actually know what I want, but I didn´t want to explain that much. Basicly I want to prigram the game in a way that something: an accident, happens always in the same place, at teh same time. And I also want some participants to have restrictions (example, some parts of the map are not visible for them).
Anyways, i will digging more.
Thanks again

**Comment by ldpl** [comment]
OpenTTD multiplayer is designed in a way that every client has complete game state and executes all actions in same order to keep that state in sync between clients. To implement limited visibility properly you'll need to find some way to make networking work without the complete game state which, depending on how much you want to hide may mean rewriting the whole game engine. Alternatively, you can assume that clients play fair and let them keep the whole state, but just hide whatever you want visually. Either way this is not something game scripts are capable of doing and would require a lot of changes to the game itself (patching). 
With disasters/accidents it's a bit more simple as implementing them doesn't contradict game design but the current script API doesn't cover that area well. So depending on what exactly you want your "accidents" to do you may still need to implement some new methods for the script api, meaning patching the game code itself.
So, in short, be prepared that current modding API of the game is very much unsuited for implementing your ideas. And, while it’s not entirely impossible, I wouldn’t advise someone without programming experience to take on such a complex task.

**Reply by lgaleano07** [reply]
Thank you so much. I also think I can find a way "a more mechanical one", to manipulate the game session, because is not an easy thing to do.
Thanks

**Comment by LordAro** [comment]
The TrAIns AI was written for the author's Masters thesis, as I recall - it may be of interest to you - https://www.luisrios.eti.br/public/en_us/research/trains/

**Reply by lgaleano07** [reply]
thanks! I will take a look at it
